### PR TITLE
AddChain inbound CCIP integration test

### PR DESCRIPTION
**Pull request by connorwstein**
[CCIP-3062](https://smartcontract-it.atlassian.net/browse/CCIP-3062)

- Integrates the MCMS library for adding inbound lanes to a new chain. The test generates a proposal to enable a 4th chain while there are 3 pre-existing chains, signs and executes it, then sends traffic via the test router for the new lanes. Serves an example of how to use the MCMS library for future workflows. 
- Renaming PR -> FeeQuoter
- Rename to OnRamp/OffRamp
- Add test router deployment and dummy MCMS key
- Extracting some test utilities

Follow ups:
- Several TODOs around parameterizing MCMS inputs https://smartcontract-it.atlassian.net/browse/CCIP-3411
- Enabling outbound from the new chain (2nd half of adding a new chain) https://smartcontract-it.atlassian.net/browse/CCIP-3412

[CCIP-3062]: https://smartcontract-it.atlassian.net/browse/CCIP-3062?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: integration-tests/deployment/ccip/add_chain.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+		enableOnRampDest, err := state.Chains[source].OnRamp.ApplyDestChainConfigUpdates(SimTransactOpts(), []onramp.OnRampDestChainConfigArgs{
+			{
+				DestChainSelector: newChainSel,
+				Router:            state.Chains[source].TestRouter.Address(),
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+		enablePriceRegDest, err := state.Chains[source].FeeQuoter.ApplyDestChainConfigUpdates(
```
Comment (AnieeG) [inline integration-tests/deployment/ccip/add_chain.go]: can we use feequoter reference all naming

File: integration-tests/deployment/ccip/add_chain.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+				},
+				{
+					// Set initial dest prices to unblock testing.
+					To:    state.Chains[source].FeeQuoter.Address(),
+					Data:  initialPrices.Data(),
+					Value: big.NewInt(0),
+				},
+				{
+					To:    state.Chains[source].FeeQuoter.Address(),
+					Data:  enablePriceRegDest.Data(),
```
Comment (AnieeG) [inline integration-tests/deployment/ccip/add_chain.go]: enableFeeQuoterDest

File: integration-tests/deployment/ccip/deploy.go [hunk 3]
```diff
@@ -378,7 +427,7 @@ func DeployChainContracts(e deployment.Environment, chain deployment.Chain, ab d
 		return ab, err
 	}
 
-	feeQuoter, err := deployContract(e.Logger, chain, ab,
+	priceRegistry, err := deployContract(e.Logger, chain, ab,
```
Comment (AnieeG) [inline integration-tests/deployment/ccip/deploy.go]: why priceRegistry?

File: integration-tests/deployment/ccip/deploy.go [hunk 4]
```diff
@@ -425,7 +474,7 @@ func DeployChainContracts(e deployment.Environment, chain deployment.Chain, ab d
 					TokenAdminRegistry: tokenAdminRegistry.Address,
 				},
 				onramp.OnRampDynamicConfig{
-					FeeQuoter:     feeQuoter.Address,
+					FeeQuoter:     priceRegistry.Address,
```
Comment (AnieeG) [inline integration-tests/deployment/ccip/deploy.go]: is it not FeeQuoter?

**Review by AnieeG — COMMENTED** [review]
There are FeeQuoter and PriceRegistry references used alternatively, can we make it uniform?

File: integration-tests/deployment/ccip/add_chain.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+			},
+			{
+				To:    state.Chains[homeChainSel].CapabilityRegistry.Address(),
+				Data:  addDON.Data(),
+				Value: big.NewInt(0),
+			},
+		},
+	})
+	return timelock.NewMCMSWithTimelockProposal(
+		"1",
```
Comment (AnieeG) [inline integration-tests/deployment/ccip/add_chain.go:160]: need a TODO to have proper values populated here?

File: integration-tests/deployment/ccip/changeset/2_initial_deploy.go [hunk 6]
```diff
@@ -14,18 +16,14 @@ func Apply0002(env deployment.Environment, c ccipdeployment.DeployCCIPContractCo
...
 	if err != nil {
 		env.Logger.Errorw("Failed to deploy CCIP contracts", "err", err, "addresses", ab)
-		return deployment.ChangesetOutput{}, err
+		return deployment.ChangesetOutput{}, deployment.MaybeDataErr(err)
 	}
 	js, err := ccipdeployment.NewCCIPJobSpecs(env.NodeIDs, env.Offchain)
 	if err != nil {
 		return deployment.ChangesetOutput{}, err
 	}
-	proposal, err := ccipdeployment.GenerateAcceptOwnershipProposal(env, env.AllChainSelectors(), ab)
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/changeset/2_initial_deploy.go:23]: This can be its own changeset, probably too much to deploy all the contracts and accept ownership in the same one

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 7]
```diff
@@ -177,32 +183,26 @@ func AddChainConfig(
...
 	if _, err := deployment.ConfirmIfNoError(h, tx, err); err != nil {
 		return ccip_config.CCIPConfigTypesChainConfigInfo{}, err
 	}
 	lggr.Infow("Applied chain config updates", "chainConfig", chainConfig)
 	return chainConfig, nil
 }
 
-func AddDON(
+// AddDONToCR adds a DON with the given nodes to the capabilities registry.
+func BuildAddDONArgs(
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/deploy_home_chain.go:192]: Need helpers for constructing the args in the case where we want to use SimTransactOpts to build proposal data

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 8]
```diff
@@ -298,33 +298,95 @@ func AddDON(
...
-		return err
+		return nil, err
 	}
 
 	// Trim first four bytes to remove function selector.
 	encodedConfigs := encodedCall[4:]
+	return encodedConfigs, nil
+}
+
+func LatestCCIPDON(registry *capabilities_registry.CapabilitiesRegistry) (*capabilities_registry.CapabilitiesRegistryDONInfo, error) {
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/deploy_home_chain.go:305]: Parsing events is more annoying when the don is added as part of proposal, so using this to get the donID in the case where you know precisely one don was added.

File: integration-tests/deployment/ccip/deploy.go [hunk 9]
```diff
@@ -253,16 +261,37 @@ func DeployChainContracts(e deployment.Environment, chain deployment.Chain, ab d
...
+	publicKey := TestXXXMCMSSigner.Public().(*ecdsa.PublicKey)
+	// Convert the public key to an Ethereum address
+	address := crypto.PubkeyToAddress(*publicKey)
+	c, err := configwrappers.NewConfig(1, []common.Address{address}, []configwrappers.Config{})
+	if err != nil {
+		e.Logger.Errorw("Failed to create config", "err", err)
+		return ab, err
+	}
+	groupQuorums, groupParents, signerAddresses, signerGroups := c.ExtractSetConfigInputs()
+	mcmsTx, err := mcm.Contract.SetConfig(chain.DeployerKey,
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/deploy.go:275]: Just a 1of1 with the test key for now. Need to parameterize in a follow up.

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 10]
```diff
@@ -52,6 +52,21 @@ const (
 	MaxDurationShouldTransmitAcceptedReport = 10 * time.Second
 )
 
+var (
+	// TODO: sort out why the mismatch here
+	//CCIPCapabilityId [32]byte = utils.MustHash(hexutil.Encode(MustABIEncode(`[{"type": "string"}, {"type": "string"}]`, CapabilityLabelledName, CapabilityVersion)))
+	CCIPCapabilityId = MustHashFromBytes(hexutil.MustDecode("0xe0da3c2b9005178f4731c9f40164f1933ad00bac9d6c13ad4ca1a8a763416380"))
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/deploy_home_chain.go]: This just avoids RPC calls all the time to get the capability ID. We expect it to be fixed per version

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 11]
```diff
@@ -164,10 +172,8 @@ func AddChainConfig(
 	ccipConfig *ccip_config.CCIPConfig,
 	chainSelector uint64,
 	p2pIDs [][32]byte,
-	f uint8,
 ) (ccip_config.CCIPConfigTypesChainConfigInfo, error) {
 	// Need to sort, otherwise _checkIsValidUniqueSubset onChain will fail
-	sortP2PIDS(p2pIDs)
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/deploy_home_chain.go:170]: sorting these once in the helper

File: integration-tests/deployment/environment.go [hunk 12]
```diff
@@ -114,9 +135,16 @@ func (n Nodes) PeerIDs(chainSel uint64) [][32]byte {
 		// Might make sense to change proto as peerID is 1-1 with node?
 		peerIDs = append(peerIDs, cfg.PeerID)
 	}
+	sort.Slice(peerIDs, func(i, j int) bool {
+		return bytes.Compare(peerIDs[i][:], peerIDs[j][:]) < 0
+	})
 	return peerIDs
 }
 
+func (n Nodes) DefaultF() uint8 {
```
Comment (AnieeG) [inline integration-tests/deployment/environment.go:152]: This should exclude bootstrap right?

File: integration-tests/deployment/ccip/propose.go [hunk 13]
```diff
@@ -17,48 +23,160 @@ import (
...
+			},
+		},
+	})
+	return timelock.NewMCMSWithTimelockProposal(
+		"1",
+		2004259681, // TODO
+		[]mcms.Signature{},
+		false,
+		metaDataPerChain,
+		"blah", // TODO
```
Comment (AnieeG) [inline integration-tests/deployment/ccip/propose.go:179]: nit : ticket?

File: integration-tests/deployment/ccip/deploy.go [hunk 14]
```diff
@@ -153,46 +176,27 @@ func DeployCCIPContracts(e deployment.Environment, c DeployCCIPContractConfig) (
...
 			c.Chains[c.HomeChainSel].CapabilityRegistry,
 			c.Chains[c.HomeChainSel].CCIPConfig,
-			chainState.EvmOffRampV160,
+			chainState.OffRamp,
 			chain,
 			e.Chains[c.HomeChainSel],
-			uint8(len(nodes)/3),
-			nodes.BootstrapPeerIDs(chain.Selector)[0],
-			nodes.PeerIDs(chain.Selector),
 			nodes,
```
Comment (AnieeG) [inline integration-tests/deployment/ccip/deploy.go]: Can you change this send plugin nodes only? with real JD if it includes bootstrap it will fail for having blank OCR2 keys and transmitter account

File: integration-tests/deployment/ccip/add_chain.go (hunk 5 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/add_chain.go:160]: yeah should be part of CCIP-3411

File: integration-tests/deployment/ccip/propose.go (hunk 13 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/propose.go:179]: part of CCIP-3411 too

**Review by AnieeG — DISMISSED** [review]

File: integration-tests/deployment/environment.go [hunk 15]
```diff
@@ -150,20 +196,33 @@ func MustPeerIDFromString(s string) p2pkey.PeerID {
...
+		var multiAddr string
 		for _, chainConfig := range nodeChainConfigs.ChainConfigs {
 			if chainConfig.Chain.Type == nodev1.ChainType_CHAIN_TYPE_SOLANA {
 				// Note supported for CCIP yet.
 				continue
 			}
+			// NOTE: Assume same peerID/multiAddr for all chains.
+			// Might make sense to change proto as peerID/multiAddr is 1-1 with nodeID?
+			peerID = MustPeerIDFromString(chainConfig.Ocr2Config.P2PKeyBundle.PeerId)
+			multiAddr = chainConfig.Ocr2Config.Multiaddr
```
Comment (AnieeG) [inline integration-tests/deployment/environment.go:219]: The multiaddr is only populated for bootstrap

File: integration-tests/deployment/ccip/deploy_home_chain.go (hunk 11 above)
Comment (asoliman92) [reply integration-tests/deployment/ccip/deploy_home_chain.go:170]: Nice!

File: integration-tests/deployment/environment.go (hunk 15 above)
Comment (asoliman92) [reply integration-tests/deployment/environment.go:219]: Out of curiosity, will it cause issues if it was set for non bootstrap nodes?

File: integration-tests/deployment/ccip/add_chain.go [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+	homeChainSel uint64,
+	newChainSel uint64,
+	sources []uint64,
+) (*timelock.MCMSWithTimelockProposal, error) {
+	// Generate proposal which enables new destination (from test router) on all source chains.
+	var batches []timelock.BatchChainOperation
+	metaDataPerChain := make(map[mcms.ChainIdentifier]timelock.MCMSWithTimelockChainMetadata)
+	for _, source := range sources {
+		chain, _ := chainsel.ChainBySelector(source)
+		enableOnRampDest, err := state.Chains[source].OnRamp.ApplyDestChainConfigUpdates(SimTransactOpts(), []onramp.OnRampDestChainConfigArgs{
```
Comment (makramkd) [inline integration-tests/deployment/ccip/add_chain.go:33]: Does this actually "send" the tx to the simchain? Proposals just need calldata right? I'm wondering what'll happen here for the non-sim chains which have "automine" or similar on.

File: integration-tests/deployment/ccip/add_chain.go [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+			},
+			{
+				To:    state.Chains[homeChainSel].CapabilityRegistry.Address(),
+				Data:  addDON.Data(),
+				Value: big.NewInt(0),
+			},
+		},
+	})
+	return timelock.NewMCMSWithTimelockProposal(
+		"1",
```
Comment (makramkd) [inline integration-tests/deployment/ccip/add_chain.go:160]: Comments on string/bool/int literal fields would be useful

File: integration-tests/deployment/ccip/add_chain_test.go [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+	//  Deploy contracts to new chain
+	newAddresses, err := DeployChainContracts(e.Env, e.Env.Chains[newChain], deployment.NewMemoryAddressBook())
+	require.NoError(t, err)
+	require.NoError(t, e.Ab.Merge(newAddresses))
+	state, err = LoadOnchainState(e.Env, e.Ab)
+	require.NoError(t, err)
+
+	// Transfer onramp/fq ownership to timelock.
+	// Enable the new dest on the test router.
+	for _, source := range initialDeploy {
```
Comment (makramkd) [inline integration-tests/deployment/ccip/add_chain_test.go:56]: Semantically distinct code blocks below (already designated by comments) can probably be moved to functions to make this test read/understand better

File: integration-tests/deployment/ccip/add_lane_test.go [hunk 19]
```diff
@@ -1 +1,57 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+// TestAddLane covers the workflow of adding a lane
+// between existing supported chains in CCIP.
+func TestAddLane(t *testing.T) {
+	// TODO: The offchain code doesn't yet support partial lane
+	// enablement, need to address then re-enable this test.
```
Comment (makramkd) [inline integration-tests/deployment/ccip/add_lane_test.go:15]: What is needed still for offchain?

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 20]
```diff
@@ -52,6 +52,21 @@ const (
 	MaxDurationShouldTransmitAcceptedReport = 10 * time.Second
 )
 
+var (
+	// TODO: sort out why the mismatch here
+	//CCIPCapabilityId [32]byte = utils.MustHash(hexutil.Encode(MustABIEncode(`[{"type": "string"}, {"type": "string"}]`, CapabilityLabelledName, CapabilityVersion)))
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy_home_chain.go]: You shouldn't hash the hex encoded string, just the ABI-encoded bytes

File: integration-tests/deployment/ccip/deploy_home_chain.go (hunk 10 above)
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy_home_chain.go]:
```suggestion
	CCIPCapabilityID = MustHashFromBytes(hexutil.MustDecode("0xe0da3c2b9005178f4731c9f40164f1933ad00bac9d6c13ad4ca1a8a763416380"))
```

File: integration-tests/deployment/ccip/deploy_home_chain.go (hunk 20 above)
Comment (makramkd) [reply integration-tests/deployment/ccip/deploy_home_chain.go]: See https://github.com/smartcontractkit/chainlink/blob/a2514d725f66e2c4a66d9bd0dd211a617fa32367/core/capabilities/ccip/common/common.go#L12-L23

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 21]
```diff
@@ -52,6 +52,21 @@ const (
...
+	CCIPCapabilityId = MustHashFromBytes(hexutil.MustDecode("0xe0da3c2b9005178f4731c9f40164f1933ad00bac9d6c13ad4ca1a8a763416380"))
+)
+
+func MustHashFromBytes(b []byte) [32]byte {
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy_home_chain.go]:
```suggestion
func MustBytesToBytes32(b []byte) [32]byte {
```

Might be a better name since this doesn't actually hash anything

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 22]
```diff
@@ -115,27 +131,19 @@ func DeployCapReg(lggr logger.Logger, chains map[uint64]deployment.Chain, chainS
...
-	})
-}
-
 func AddNodes(
 	capReg *capabilities_registry.CapabilitiesRegistry,
 	chain deployment.Chain,
 	p2pIDs [][32]byte,
-	capabilityIDs [][32]byte,
 ) error {
 	// Need to sort, otherwise _checkIsValidUniqueSubset onChain will fail
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy_home_chain.go]: Can remove this comment, sorting no longer needed

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 23]
```diff
@@ -164,10 +172,8 @@ func AddChainConfig(
 	ccipConfig *ccip_config.CCIPConfig,
 	chainSelector uint64,
 	p2pIDs [][32]byte,
-	f uint8,
 ) (ccip_config.CCIPConfigTypesChainConfigInfo, error) {
 	// Need to sort, otherwise _checkIsValidUniqueSubset onChain will fail
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy_home_chain.go]: Can remove this comment, sorting no longer needed

File: integration-tests/deployment/ccip/deploy_home_chain.go (hunk 11 above)
Comment (makramkd) [reply integration-tests/deployment/ccip/deploy_home_chain.go:170]: Sorting no longer needed, can remove it entirely

File: integration-tests/deployment/ccip/add_chain.go (hunk 16 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/add_chain.go:33]: it doesn't - uses the `NoSend` flag in transact opts, just a scalable way to get the calldata for proposals

File: integration-tests/deployment/ccip/add_lane_test.go (hunk 19 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/add_lane_test.go:15]: nothing now if the IsEnabled flag work is done, then we can complete this test

File: integration-tests/deployment/ccip/add_chain.go (hunk 17 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/add_chain.go:160]: yeah lets punt to CCIP-3411

File: integration-tests/deployment/ccip/add_chain_test.go (hunk 18 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/add_chain_test.go:56]: agree - was thinking we wait for another proposal to see if there are nice re-usable utilities we can extract

File: integration-tests/deployment/ccip/add_chain.go (hunk 16 above)
Comment (makramkd) [reply integration-tests/deployment/ccip/add_chain.go:33]: Oh that's clever, alright did not notice that

File: integration-tests/deployment/environment.go (hunk 15 above)
Comment (connorwstein) [reply integration-tests/deployment/environment.go:219]: I think leaving it blank for non-bootstraps is ok for now. @asoliman92 no don't think it will

**Review by asoliman92 — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]
